cluster-autoscaler/simulator/scheduler_based_predicates_checker.go

```diff
@@ -0,0 +1,339 @@
+	nodes, err := p.nodeLister.List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("could not list Nodes; %v", err)
+	}
+	pods, err := p.podLister.List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("could not list Pods; %v", err)
+	}
+	nodeInfoMap := CreateNodeInfoMap(pods, nodes)
+	newSnapshot := NewSnapshot(nodeInfoMap)
+	p.delegatingSharedLister.UpdateDelegate(newSnapshot)
+	return nil
+}
+
+// FitsAnyNode checks if the given pod can be place on any of the given nodes.
+func (p *SchedulerBasedPredicateChecker) FitsAnyNode(clusterSnapshot ClusterSnapshot, pod *apiv1.Pod, nodeInfos map[string]*scheduler_nodeinfo.NodeInfo) (string, error) {
+	if clusterSnapshot != nil {
+		if nodeInfos != nil {
+			klog.Errorf("clusterSnapshot and nodeInfos are mutually exclusive!!!!")
+		}
+		return p.fitsAnyNode(clusterSnapshot, pod)
+	}
+	return p.fitsAnyNodeDeprecated(pod, nodeInfos)
+}
+
+// FitsAnyNode checks if the given pod can be place on any of the given nodes.
+func (p *SchedulerBasedPredicateChecker) fitsAnyNodeDeprecated(pod *apiv1.Pod, nodeInfos map[string]*scheduler_nodeinfo.NodeInfo) (string, error) {
+	p.delegatingSharedLister.UpdateDelegate(p.informerBasedShardLister)
+	defer p.delegatingSharedLister.ResetDelegate()
+
+	state := scheduler_framework.NewCycleState()
+	preFilterStatus := p.framework.RunPreFilterPlugins(context.TODO(), state, pod)
+	if !preFilterStatus.IsSuccess() {
+		return "", fmt.Errorf("error running pre filter plugins for pod %s; %s", pod.Name, preFilterStatus.Message())
+	}
+
+	for _, nodeInfo := range nodeInfos {
+		// Be sure that the node is schedulable.
+		if nodeInfo.Node().Spec.Unschedulable {
+			continue
+		}
+
+		filterStatuses := p.framework.RunFilterPlugins(context.TODO(), state, pod, nodeInfo)
+		ok := true
+		for _, filterStatus := range filterStatuses {
+			if !filterStatus.IsSuccess() {
+				ok = false
+				break
+			}
+		}
+		if ok {
+			return nodeInfo.Node().Name, nil
+		}
+	}
+	return "", fmt.Errorf("cannot put pod %s on any node", pod.Name)
+}
+
+func (p *SchedulerBasedPredicateChecker) fitsAnyNode(clusterSnapshot ClusterSnapshot, pod *apiv1.Pod) (string, error) {
+	var nodeInfosList []*scheduler_nodeinfo.NodeInfo
+	schedulerLister, err := clusterSnapshot.GetSchedulerLister()
+	if err != nil {
+		// TODO(scheduler_framework_integration) distinguish from internal error and predicate error
+		klog.Errorf("Error obtaining SharedLister from clusterSnapshot")
+		return "", fmt.Errorf("error obtaining SharedLister from clusterSnapshot")
+	}
+	nodeInfosList, err = schedulerLister.NodeInfos().List()
+	if err != nil {
+		// TODO(scheduler_framework_integration) distinguish from internal error and predicate error
+		klog.Errorf("Error obtaining nodeInfos from schedulerLister")
+		return "", fmt.Errorf("error obtaining nodeInfos from schedulerLister")
+	}
+	p.delegatingSharedLister.UpdateDelegate(schedulerLister)
+	defer p.delegatingSharedLister.ResetDelegate()
```

Comment:
Maybe trying to use an unset delegate should just blow up more explicitly and not pass empty snapshot.

Reply:
Yeah - I guess - having empty delegate by default was mostly for intermediate state. When not all the code was migrated. And I wanted CA to "somewhat" work. Will change it so it fails if delegate is not set.